Modules/_sqlite/connection.c

```diff
@@ -1817,6 +1817,117 @@ pysqlite_connection_create_collation_impl(pysqlite_Connection *self,
     Py_RETURN_NONE;
 }
 
+#ifdef PY_SQLITE_HAVE_SERIALIZE
+/*[clinic input]
+_sqlite3.Connection.serialize as serialize
+
+    *
+    name: str = "main"
+        Which database to serialize.
+
+Serialize a database into a byte string. Non-standard.
+
+For an ordinary on-disk database file, the serialization is just a copy of the
+disk file. For an in-memory database or a "temp" database, the serialization is
+the same sequence of bytes which would be written to disk if that database
+where backed up to disk.
+[clinic start generated code]*/
+
+static PyObject *
+serialize_impl(pysqlite_Connection *self, const char *name)
+/*[clinic end generated code: output=97342b0e55239dd3 input=4b6efe5a4d524470]*/
+{
+    if (!pysqlite_check_thread(self) || !pysqlite_check_connection(self)) {
+        return NULL;
+    }
+
+    sqlite3_int64 size;
+    const unsigned int flags = 0;
+    const char *data;
+
+    Py_BEGIN_ALLOW_THREADS
+    data = (const char *)sqlite3_serialize(self->db, name, &size, flags);
+    Py_END_ALLOW_THREADS
+
+    if (data == NULL) {
+        PyErr_Format(self->OperationalError, "unable to serialize '%s'",
+                     name);
+        return NULL;
+    }
+    PyObject *res = PyBytes_FromStringAndSize(data, size);
+    sqlite3_free((void *)data);
+    return res;
+}
+
+/*[clinic input]
+_sqlite3.Connection.deserialize as deserialize
+
+    data: Py_buffer(accept={buffer, str})
+        The serialized database content.
+    /
+    *
+    name: str = "main"
+        Which database to reopen with the deserialization.
+
+Load a serialized database. Non-standard.
+
+The deserialize interface causes the database connection to disconnect from the
+target database, and then reopen it as an in-memory database based on the given
+serialized data.
+
+The deserialize interface will fail with SQLITE_BUSY if the database is
+currently in a read transaction or is involved in a backup operation.
+[clinic start generated code]*/
+
+static PyObject *
+deserialize_impl(pysqlite_Connection *self, Py_buffer *data,
+                 const char *name)
+/*[clinic end generated code: output=e394c798b98bad89 input=7cf51653d0c40e22]*/
+{
+    if (!pysqlite_check_thread(self) || !pysqlite_check_connection(self)) {
+        return NULL;
+    }
+
+    /* Transfer ownership of the buffer to SQLite:
+     * - Move buffer from Py to SQLite
+     * - Tell SQLite to free buffer memory
+     * - Tell SQLite that it is permitted to grow the resulting database
+     *
+     * Make sure we don't overflow sqlite3_deserialize(); it accepts a signed
+     * 64-bit int as its data size argument, but sqlite3_malloc64 accepts an
+     * unsigned 64-bit int as its size argument.
+     *
+     * We can safely use sqlite3_malloc64 here, since it was introduced before
+     * the serialize APIs.
+     */
+    if (data->len > 9223372036854775807) {
```

Comment:
```suggestion
    if (data->len >= (2 << 62)) {
```
Or some other variation that's not just a magic number.

Reply:
Seems like this broke the Windows builds; I think just writing it out literally may be the best option, to avoid compiler tricks. There also seems to be some precedent for writing these out explicitly:

```
$ grep -r 9223372036854775807 * | wc -l
      27
$ grep -r 9223372036854775808 * | wc -l
      19
$ grep -r "2 << 62" * | wc -l          
       1
```

Reply:
I added a comment instead; it seems that the MSVC compiler does some tricks here (I tried to find out via godbolt.org, but to no avail; perhaps explicitly marking the shifted literal as `long long` would have worked...) The code base also seem to prefer the explicit to the implicit here:

```
$ grep -r "922337203685477580[78]" * | wc -l
      47
$ grep -r "[12] << 6[23]" * | wc -l          
      13
```

Reply:
On MSVC, `(1 << 63) - 1` triggers [MSVC compiler warning C4293](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-1-c4293?view=msvc-170). Using `(1LL << 63) - 1` seems to fix it, but the MS docs promotes casting to `unsigned __int64`.

I prefer just writing it out instead; also, there is already a comment explaining why we do this :)

```
     * Make sure we don't overflow sqlite3_deserialize(); it accepts a signed
     * 64-bit int as its data size argument, but sqlite3_malloc64 accepts an
     * unsigned 64-bit int as its size argument.
```

Reply:
(Actually, I can reduce that comment; only the first part is important.)